### PR TITLE
[Program Overview Page] Add new application steps fields

### LINKED ISSUE #9086: [Program Overview Page] Add new application steps fields to program create/edit page
Allow admins to add up to five application steps describing the application process.

Acceptance criteria:

- [x]  The program create/edit page includes space to add up titles and descriptions for up to five application steps
- [x] These fields are grouped together in a new "How to apply" section that is below the "Program overview" section
- [x] The first application step is required
- [x] When the admin saves the program, the application steps are saved to the `application_steps` column on the `programs` table as an array of application steps that includes json representation of both the title and description as `LocalizedStrings`
- [x] If an admin attempts to save the program without filling out the first application step, they are shown an error saying that the first application step is required
- [x] If an admin attempts to save the program with an application step that is missing either the title or description, they are shown an error saying that application steps must have both a title and a description
- [x] If an admin leaves a step in the middle blank, the other steps are condensed so there is no blank step in the middle

See [the TDD](https://docs.google.com/document/d/1rP4kqltv122Wfp9-maJlvm5rzFkFcWrCt5w0lKRoZEs/edit?tab=t.0#heading=h.rto1cqcwpkdl) for more information about the implementation plan and sample code

### PR DESCRIPTION
**Pull request by rockycodes**
### Description

This PR adds new "application step" fields to the admin program create/edit form which are saved to the db when the admin submits the form. At least one application step is required per program and admin can enter up to five steps. Each application step contains a localized title and localized description.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [x] Extended the README / documentation, if necessary. For user-facing features, consider updating [the user docs](https://github.com/civiform/docs). For "under-the-hood" changes or things more relevant to developers, consider updating [the dev wiki](https://github.com/civiform/civiform/wiki).
- [x] Ensured PII wasn't added to any new logs, unless it was guarded by `isDevOrStaging`

### Instructions for manual testing

- Create a new program, add an application step, confirm it submits correctly and shows up when you re-open the program
- Edit an existing program, add an application step, confirm it submits correctly and shows up when you re-open the program
- Attempt to submit the form without adding any application steps and see the error
- Attempt to submit the form with an application step that is missing either its title or description and see the error
- Submit the form with application steps that are blank and confirm blank steps are ignored. If blank steps are in the middle, the filled in steps should collapse (Eg. if step 2 is blank, but steps 1 and 3 are filled in, the data for step 3 will be saved as step 2)

### Issue(s) this completes

Fixes #9086

### CODE REVIEW

File: server/test/services/applicant/ApplicantServiceFastForwardEnabledTest.java [hunk 1]
```diff
@@ -3163,6 +3164,8 @@ public void maybeEligibleProgramsForApplicant_doesNotIncludeIneligibleSubmittedA
                     .setEligibilityIsGating(false)
                     .setAcls(new ProgramAcls())
                     .setCategories(ImmutableList.of())
+                    .setApplicationSteps(
```
Comment (rockycodes) [inline server/test/services/applicant/ApplicantServiceFastForwardEnabledTest.java:3167]: For some reason these tests aren't failing without a short description, which is also required. I'm not sure why they require application steps but not description.

File: server/test/controllers/admin/AdminProgramControllerTest.java [hunk 2]
```diff
@@ -946,4 +1013,52 @@ public void publishProgram_nonDraftProgram_throwsException() throws Exception {
     assertThatThrownBy(() -> controller.publishProgram(fakeRequest(), program.id))
         .isInstanceOf(NotChangeableException.class);
   }
+
+  @Test
+  public void buildApplicationSteps_transformsDataIntoApplicationStepObjects() {
```
Comment (rockycodes) [inline server/test/controllers/admin/AdminProgramControllerTest.java:1018]: Here's a new test I added.

File: server/test/services/applicant/ApplicantServiceFastForwardEnabledTest.java (hunk 1 above)
Comment (melanie-exygy) [reply server/test/services/applicant/ApplicantServiceFastForwardEnabledTest.java:3167]: Hm, I can't seem to find an explanation.

File: server/test/services/program/ProgramServiceTest.java [hunk 3]
```diff
@@ -843,11 +871,43 @@ public void validateProgramDataForCreate_returnsNoErrorsInSELECT_TIMode() {
...
+    ImmutableList<ApplicationStep> applicationSteps =
+        ImmutableList.of(
+            new ApplicationStep("title one", "description one"),
+            new ApplicationStep("title two", ""),
+            new ApplicationStep("", "description three"));
+    ImmutableSet<CiviFormError> errors =
+        ps.checkApplicationStepErrors(errorsBuilder, applicationSteps).build();
+
+    assertThat(errors.size()).isEqualTo(2);
+    assertThat(errors.contains(CiviFormError.of("Application step 2 is missing a description")))
```
Comment (melanie-exygy) [inline server/test/services/program/ProgramServiceTest.java:906]: **praise:** Thanks for adding these tests!

**Review by avaleske — APPROVED** [review]
lgtm after melanie's comments!

Also out of curiosity, and out of scope for this: How hard would it be to hide the 2-5th divs behind a button so the program creation screen is less cluttered?

**Review by melanie-exygy — APPROVED** [review]

File: server/app/views/admin/programs/ProgramFormBuilder.java [hunk 4]
```diff
@@ -328,6 +348,53 @@ private FormTag buildProgramForm(
...
+      String title = "";
+      String description = "";
+      if (i + 1 <= applicationSteps.size()) {
+        title = applicationSteps.get(i).get("title");
+        description = applicationSteps.get(i).get("description");
+      }
+      div.with(buildApplicationStepDiv(i, title, description));
+    }
+    return div;
+  }
```
[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/views/admin/programs/ProgramFormBuilder.java:365]: Good suggestion, I think that's easier to reason about. Thanks!

**Comment by rockycodes** [comment]
> lgtm after melanie's comments!
> 
> Also out of curiosity, and out of scope for this: How hard would it be to hide the 2-5th divs behind a button so the program creation screen is less cluttered?

V1 is to show all the steps at once, but V2 is to make adding/removing them dynamic! I think it won't be too hard, just require some JS. But product opted to leave that out for now in favor of getting this out as quickly as possible.

File: server/app/controllers/admin/AdminProgramController.java [hunk 5]
```diff
@@ -312,4 +327,23 @@ private Result getSaveProgramDetailsRedirect(
       return redirect(routes.AdminProgramBlocksController.index(programId).url());
     }
   }
+
+  /** Turn application step form data into ApplicationStep objects */
+  @VisibleForTesting
+  public ImmutableList<ApplicationStep> buildApplicationSteps(
```
[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/controllers/admin/AdminProgramController.java:333]: TIL that leaving off the modifier makes it package private! Thank you!

File: browser-test/src/support/admin_programs.ts [hunk 6]
```diff
@@ -188,6 +189,18 @@ export class AdminPrograms {
       confirmationMessage,
     )
 
+    for (let i = 0; i < applicationSteps.length; i++) {
+      const indexPlusOne = i + 1
+      await this.page.fill(
```
[the remark this answers is not in the document]
Comment (rockycodes) [reply browser-test/src/support/admin_programs.ts]: great idea! updated.

File: server/app/views/admin/programs/ProgramFormBuilder.java [hunk 7]
```diff
@@ -311,6 +324,13 @@ private FormTag buildProgramForm(
             .setValue(externalLink)
             .getInputTag()
             .withClass(SPACE_BETWEEN_FORM_ELEMENTS),
+        h2("How to apply").withClasses("py-2", "mt-6", "font-semibold"),
+        AlertComponent.renderSlimAlert(
+            AlertType.INFO,
+            "Application steps will be visible to applicants at a future date.",
+            false,
+            "my-2"),
+        div().with(buildApplicationSteps(applicationSteps)),
```
[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/views/admin/programs/ProgramFormBuilder.java]: i updated it per austin's suggestion so the 5 steps aren't wrapped in an extra div

**Review by avaleske — APPROVED** [review]

**Pull request merged**